service/topology_coordinator.cc

```diff
@@ -2952,6 +2994,7 @@ future<> run_topology_coordinator(
         on_fatal_internal_error(rtlogger, format("unhandled exception in topology_coordinator::run: {}", ex));
     }
     co_await lifecycle_notifier.unregister_subscriber(&coordinator);
+    co_await coordinator.stop();
```

Comment:
How long this may take?  We will not start new coordinator on this node until this completes. So suppose we lost leadership and hang here for 10 minutes, but we re-acquire the leadership after a second. The cluster will run without the coordinator for 10 minutes.

Reply:
@gleb-cloudius tl;dr, it does not take 10 minutes. on the contrary, the all the operations are aborted by the same `as` which aborts the grand loop.

long story:
----

how long it may take depends on the sum of following three time periods:

- how long `tablet_migration_state.streaming` may take to resolve:  yes, we need to wait until `storage_service::stream_tablet()` completes or gets cancelled. but we do pass a `cancellable` down to rpc when performing streaming. and this `cancellable` instance is associated with the `abort_source` which aborts the grand loop in `topology_coordinator::run()`.
- how long `cleanup` may take to resolve: this may take relatively long time. but the same applies to the tablet_cleanup rpc call.
- how long the active barrier in `barriers` may take to resolve: barrier works like a mutex in `topology_coordinator::handle_tablet_migration()`, it serializes the mutations to the shared state, like `updates`. when we done with an operation, we always check if current term ends, and if that's the case, a `term_changed_error` exception is thrown, so all operations are cancelled. in our case, we are not a leader anymore, which implies that a new terms starts. so waiting for all barriers in tablets would take at most the time to complete a raft operation. and, raft rpc is also cancelled by the above `as`. for instance, see https://github.com/scylladb/scylladb/blob/cfd6084eddf51879a5332d5b6261c7aee41b643f/raft/server.cc#L1452-L1482

Reply:
> @gleb-cloudius tl;dr, it does not take 10 minutes. on the contrary, all the operations are aborted by the same `as` which aborts the grand loop.

If all the work that we are waiting for is canceled it is fine.

> ## long story:
> 
> how long it may take depends on the sum of following three time periods:
> 
>     * how long `tablet_migration_state.streaming` may take to resolve:  yes, we need to wait until `storage_service::stream_tablet()` completes or gets cancelled. but we do pass a `cancellable` down to rpc when performing streaming. and this `cancellable` instance is associated with the `abort_source` which aborts the grand loop in `topology_coordinator::run()`.
> 
>     * how long `cleanup` may take to resolve: this may take relatively long time. but the same applies to the tablet_cleanup rpc call.
> 
Do we cancel the cleanup?

>     * how long the active barrier in `barriers` may take to resolve: barrier works like a mutex in `topology_coordinator::handle_tablet_migration()`, it serializes the mutations to the shared state, like `updates`. when we done with an operation, we always check if current term ends, and if that's the case, a `term_changed_error` exception is thrown, so all operations are cancelled. in our case, we are not a leader anymore, which implies that a new terms starts. so waiting for all barriers in tablets would take at most the time to complete a raft operation. and, raft rpc is also cancelled by the above `as`. for instance, see https://github.com/scylladb/scylladb/blob/cfd6084eddf51879a5332d5b6261c7aee41b643f/raft/server.cc#L1452-L1482

Barrier can hang for a lot of time since it waits for all previous requests to complete. But this is pre-existing since if non tablet barrier is stuck `coordinator::run` will also not return. We need to make topology rpcs cancelable if they are not yet.

Reply:
> > @gleb-cloudius tl;dr, it does not take 10 minutes. on the contrary, all the operations are aborted by the same `as` which aborts the grand loop.
> 
> If all the work that we are waiting for is canceled it is fine.
> 
> > ## long story:
> > how long it may take depends on the sum of following three time periods:
> > ```
> > * how long `tablet_migration_state.streaming` may take to resolve:  yes, we need to wait until `storage_service::stream_tablet()` completes or gets cancelled. but we do pass a `cancellable` down to rpc when performing streaming. and this `cancellable` instance is associated with the `abort_source` which aborts the grand loop in `topology_coordinator::run()`.
> > 
> > * how long `cleanup` may take to resolve: this may take relatively long time. but the same applies to the tablet_cleanup rpc call.
> > ```
> 
> Do we cancel the cleanup?

let me be more specific.

- see https://github.com/scylladb/scylladb/blob/cfd6084eddf51879a5332d5b6261c7aee41b643f/service/topology_coordinator.cc#L1246
- see https://github.com/scylladb/scylladb/blob/cfd6084eddf51879a5332d5b6261c7aee41b643f/service/topology_coordinator.cc#L1264

both the rpc calls pass `_as` down. and we are using the same machinery provided by seastar rpc  for registering rpc handler and for performing rpc calls.
> 
> > ```
> > * how long the active barrier in `barriers` may take to resolve: barrier works like a mutex in `topology_coordinator::handle_tablet_migration()`, it serializes the mutations to the shared state, like `updates`. when we done with an operation, we always check if current term ends, and if that's the case, a `term_changed_error` exception is thrown, so all operations are cancelled. in our case, we are not a leader anymore, which implies that a new terms starts. so waiting for all barriers in tablets would take at most the time to complete a raft operation. and, raft rpc is also cancelled by the above `as`. for instance, see https://github.com/scylladb/scylladb/blob/cfd6084eddf51879a5332d5b6261c7aee41b643f/raft/server.cc#L1452-L1482
> > ```
> 
> Barrier can hang for a lot of time since it waits for all previous requests to complete. But this is pre-existing since if non tablet barrier is stuck `coordinator::run` will also not return. We need to make topology rpcs cancelable if they are not yet.

i think they are cancellable.  the mutations operations share a semaphore of "1", but once the current one is aborted, all following operations waiting in the queue will be cancelled once they acquire the semaphore.